File: app/src/main/java/com/vocaby/app/models/EntryModel.java
package com.vocaby.app.models;

import android.os.Parcel;
import android.os.Parcelable;

import java.util.ArrayList;
import java.util.List;

public class EntryModel implements Parcelable {
    private int id;
    private final String entry;
    private String pronunciation;
    // TODO: Change List to Hashmap?
    private List<DefinitionGroupModel> definitionGroups;

    public EntryModel(int id, String entry) {
        this.id = id;
        this.entry = entry;
        pronunciation = "";
        definitionGroups = new ArrayList<>();
    }

    public EntryModel(String entry) {
        id = -1;
        this.entry = entry;
        pronunciation = "";
        definitionGroups = new ArrayList<>();
    }

    protected EntryModel(Parcel in) {
        definitionGroups = new ArrayList<>();

        id = in.readInt();
        entry = in.readString();
        pronunciation = in.readString();
        in.readTypedList(definitionGroups, DefinitionGroupModel.CREATOR);
    }

    public static final Creator<EntryModel> CREATOR = new Creator<EntryModel>() {
        @Override
        public EntryModel createFromParcel(Parcel in) {
            return new EntryModel(in);
        }

        @Override
        public EntryModel[] newArray(int size) {
            return new EntryModel[size];
        }
    };

    public void setPronunciation(String pronunciation) {
        this.pronunciation = pronunciation;
    }

    public String getPronunciation() { return this.pronunciation; }

    public String getEntry() { return entry; }

    public boolean isEmpty() {
        return definitionGroups.isEmpty();
    }

    public void setDefinitionGroups(List<DefinitionGroupModel> newGroups) {
        definitionGroups = newGroups;
    }

    public void addDefinitionGroup(String type) {
       definitionGroups.add(new DefinitionGroupModel(type, definitionGroups.size()));
    }

    public void addDefinitionGroup(DefinitionGroupModel definitionGroupModel) {
        definitionGroups.add(definitionGroupModel);
    }

    public void replaceDefinitionGroup(String type, DefinitionGroupModel newGroup) {
        int index = getGroupIndex(type);
        definitionGroups.set(index, newGroup);
    }

    // TODO: Override list remove
    public DefinitionGroupModel removeGroup(int position) {
        DefinitionGroupModel groupToRemove = definitionGroups.get(position);
        definitionGroups.remove(position);
        if(position < definitionGroups.size()) {
            for(int i = position; i < definitionGroups.size(); i++) {
                definitionGroups.get(i).setOrder(i);
            }
        }

        return groupToRemove;
    }

    public void addDefinition(String type, String definition, String example) {
        int index = getGroupIndex(type);
        if(index != -1) {
            DefinitionGroupModel group = definitionGroups.get(index);
            if (group != null) {
                group.addNewDefinition(definition, example);
            }
        } else {
            DefinitionGroupModel newGroup = new DefinitionGroupModel(type, definitionGroups.size());
            newGroup.addNewDefinition(definition, example);
            definitionGroups.add(newGroup);
        }
    }

    public List<DefinitionGroupModel> getDefinitionGroups() {
        return definitionGroups;
    }

    public DefinitionGroupModel getFirstGroup() {
        return definitionGroups.isEmpty() ? null : definitionGroups.get(0);
    }

    public DefinitionGroupModel getLastGroup() {
        return definitionGroups.get(definitionGroups.size()-1);
    }

    public DefinitionGroupModel getDefinitionGroup(int index) {
        return definitionGroups.get(index);
    }

    public DefinitionGroupModel getDefinitionGroup(String type) {
        for (DefinitionGroupModel group : definitionGroups) {
            if (group.getType().equals(type)) return group;
        }

        return null;
    }

    public int getId() {
        return id;
    }

    public void setId(int id) {
        this.id = id;
    }


    public boolean hasGroup(String type) {
        return getGroupIndex(type) != -1;
    }

    private int getGroupIndex(String type) {
        for(int i = 0; i < definitionGroups.size(); i++) {
            if (definitionGroups.get(i).getType().equals(type)) return i;
        }

        return -1;
    }

    @Override
    public int describeContents() {
        return 0;
    }

    @Override
    public void writeToParcel(Parcel parcel, int i) {
        parcel.writeInt(id);
        parcel.writeString(entry);
        parcel.writeString(pronunciation);
        parcel.writeTypedList(definitionGroups);
    }
}

File: app/src/main/java/com/vocaby/app/models/dictionary/EntryModel.kt
package com.vocaby.app.models.dictionary

import android.os.Parcelable
import android.os.Parcel
import android.os.Parcelable.Creator
import java.util.ArrayList

class EntryModel : Parcelable {
    var id: Int
    val entry: String
    var pronunciation: String?
    var definitionGroups: MutableList<DefinitionGroupModel>

    constructor(id: Int, entry: String, pronunciation: String) {
        this.id = id
        this.entry = entry
        this.pronunciation = pronunciation
        definitionGroups = ArrayList()
    }

    constructor(entry: String) {
        id = -1
        this.entry = entry
        pronunciation = ""
        definitionGroups = ArrayList()
    }

    private constructor(`in`: Parcel) {
        definitionGroups = ArrayList()
        id = `in`.readInt()
        entry = `in`.readString().toString()
        pronunciation = `in`.readString()

        `in`.readTypedList(definitionGroups, DefinitionGroupModel)
    }

    val isEmpty: Boolean
        get() = definitionGroups.isEmpty()


    fun addDefinitionGroup(definitionGroupModel: DefinitionGroupModel) {
        definitionGroups.add(definitionGroupModel)
    }

    fun replaceDefinitionGroup(type: String, newGroup: DefinitionGroupModel) {
        val index = getGroupIndex(type)
        definitionGroups[index] = newGroup
    }

    // TODO: Override list remove
    fun removeGroup(position: Int): DefinitionGroupModel {
        val groupToRemove = definitionGroups[position]
        definitionGroups.removeAt(position)
        if (position < definitionGroups.size) {
            for (i in position until definitionGroups.size) {
                definitionGroups[i].order = i
            }
        }
        return groupToRemove
    }

    fun addDefinition(type: String, definition: String, example: String?) {
        val index = getGroupIndex(type)
        if (index != -1) {
            val group = definitionGroups[index]
            group.addNewDefinition(definition, example)
        } else {
            val newGroup = DefinitionGroupModel(type, definitionGroups.size)
            newGroup.addNewDefinition(definition, example)
            definitionGroups.add(newGroup)
        }
    }

    val firstGroup: DefinitionGroupModel
        get() =definitionGroups[0]

    fun getDefinitionGroup(index: Int): DefinitionGroupModel {
        return definitionGroups[index]
    }

    private fun getGroupIndex(type: String): Int {
        for (i in definitionGroups.indices) {
            if (definitionGroups[i].type == type) return i
        }
        return -1
    }

    override fun describeContents(): Int {
        return 0
    }

    override fun writeToParcel(parcel: Parcel, i: Int) {
        parcel.writeInt(id)
        parcel.writeString(entry)
        parcel.writeString(pronunciation)
        parcel.writeTypedList(definitionGroups)
    }

    companion object CREATOR : Creator<EntryModel> {
        override fun createFromParcel(parcel: Parcel): EntryModel {
            return EntryModel(parcel)
        }

        override fun newArray(size: Int): Array<EntryModel?> {
            return arrayOfNulls(size)
        }
    }
}